### PR TITLE
promql.ActiveQueryTracker: Unmap mmapped file when done

### PR DESCRIPTION
**Pull request by aknuds1**
Fix `promql.ActiveQueryTracker` so that it unmaps any mmapped file in its `Close` method. Also fix `TestMMapFile` so it releases resources.

Related to https://github.com/prometheus/prometheus/pull/13948.

### CODE REVIEW

File: promql/query_logger.go [hunk 1]
```diff
@@ -108,7 +128,7 @@ func getMMapedFile(filename string, filesize int, logger log.Logger) ([]byte, io
 		return nil, nil, err
 	}
 
-	return fileAsBytes, file, err
+	return fileAsBytes, &mmapedFile{f: file, m: fileAsBytes}, err
```
Comment (machine424) [inline promql/query_logger.go:131]: I think Ben's suggestion here https://github.com/prometheus/prometheus/pull/13948#discussion_r1577884572
was that we can even close the file descriptor in `getMMapedFile` (thus we no longer need to return it)
Did you try that? Doesn't it work on windows?

**Review by machine424 — COMMENTED** [review]
Thanks!

File: promql/query_logger.go (hunk 1 above)
Comment (aknuds1) [reply promql/query_logger.go:131]: I don't have Windows, so I'm not able to test there myself. I think Ben only found documentation for Linux though, not Unix in general. Are you able to find corresponding documentation for other Unices?

Comment (aknuds1) [reply promql/query_logger.go:131]: Reading [documentation](https://learn.microsoft.com/en-us/windows/win32/memory/creating-a-file-mapping-object) on mmapping on Windows, it looks as if one should keep the file open for the lifetime of the mapping:

> In addition, the file handle should remain open until the process no longer needs the file mapping object.

It also says the following about closing the mmapped file:

> When each process finishes using the file mapping object and has unmapped all views, it must close the file mapping object's handle and the file on disk by calling CloseHandle.

**Comment by aknuds1** [comment]
@machine424 I pushed a revision also to fix flakiness in `TestQueryConcurrency`, since non-deterministic goroutine behaviour in it would cause panics in combination with my changes in this PR.

File: promql/query_logger.go [hunk 2]
```diff
@@ -83,6 +86,23 @@ func logUnfinishedQueries(filename string, filesize int, logger log.Logger) {
...
 }
 
+type mmapedFile struct {
+	f io.Closer
+	m mmap.MMap
+}
+
+func (f *mmapedFile) Close() error {
+	err := f.m.Unmap()
+	if fErr := f.f.Close(); fErr != nil && err == nil {
```
Comment (machine424) [inline promql/query_logger.go:96]: we could use two err variables and `errors.Join` them (nil is ignored)

File: promql/query_logger_test.go [hunk 3]
```diff
@@ -104,26 +105,26 @@ func TestIndexReuse(t *testing.T) {
...
+	fileAsBytes, closer, err := getMMapedFile(fpath, 2, nil)
 	require.NoError(t, err)
-	copy(fileAsBytes, "ab")
+	copy(fileAsBytes, data)
+	require.NoError(t, closer.Close())
 
-	f, err := os.Open(filename)
+	f, err := os.Open(fpath)
 	require.NoError(t, err)
+	t.Cleanup(func() {
```
Comment (machine424) [inline promql/query_logger_test.go:119]: nit: we can also `t.Cleanup(f.Close)`

File: promql/query_logger.go (hunk 1 above)
Comment (machine424) [reply promql/query_logger.go:131]: I was implicitly suggesting to give it a try on the windows job in the CI (You could still give it a try to confirm the statements in the doc if you want)

> Based on the above, I believe we should keep the current behaviour (of the PR). I also don't see that we gain much by closing the file early?

Just for less/simpler code, as we are now getting closer and closer to https://github.com/prometheus/prometheus/blob/1081e336a0c4004b6f817c70f0ff4f3810db80fa/tsdb/fileutil/mmap.go#L21, we’ll see if we can merge them in the future.

**Review by machine424 — APPROVED** [review]
lgtm, thanks.
just some nits and ideas for the future :)

File: promql/query_logger_test.go (hunk 3 above)
Comment (aknuds1) [reply promql/query_logger_test.go:119]: `f.Close()` returns an error, so that doesn't work.

File: promql/query_logger.go (hunk 2 above)
Comment (aknuds1) [reply promql/query_logger.go]: Nice suggestion, thanks!

File: promql/query_logger_test.go (hunk 3 above)
Comment (aknuds1) [reply promql/query_logger_test.go:119]: My thinking for ignoring the error is that an OS error should be outside of our control. Does it matter to the person running the tests if a file should fail to close after tests are done running?

File: promql/query_logger.go (hunk 1 above)
Comment (aknuds1) [reply promql/query_logger.go:132]: My take is that even if it would work in CI, we shouldn't go against what the Windows mmap documentation specifies.

Comment (aknuds1) [reply promql/query_logger.go:132]: I did a test with CI, and closing the file before the mmap handle actually does fail: `mmapedFile: unmapping: FlushFileBuffers: The handle is invalid.`.

File: promql/query_logger_test.go (hunk 3 above)
Comment (aknuds1) [reply promql/query_logger_test.go:119]: Hope you don't mind if I merge in spite of this one unresolved comment. I really don't think there's anything to gain from asserting on releasing an OS resource :)

Comment (machine424) [reply promql/query_logger_test.go:119]: Actually, I though f was a `queryTracker`.

**Pull request merged**